### PR TITLE
complex support for tf.math.argmin and tf.math.argmax

### PR DESCRIPTION
**Pull request by jstzwj**
This PR tries to address the issue raised in #41437
tf.math.argmin and tf.math.argmax don't support complex
The issue was that there aren't kLt, kGt, kLe and kGe ComparisonDirection in opcode kCompare of complex. Besides, MinValue and MaxValue in xla/literal_util.cc didn't support the minimum and maximum value for complex.
This PR adds complex support for tf.math.argmin and tf.math.argmax.

### CODE REVIEW

**Comment by yongtang** [comment]
In numpy, it looks like np.argmax is defined as the lexi order of the tuple `(real_part, Imag_part)` based on https://github.com/numpy/numpy/issues/10469

The following verifies:
```
>>> import numpy as np
>>> np.argmax([1+0.5j, 1-0.5j])
0
>>> np.argmax([1-0.5j, 1+0.5j])
1
```

Should the behavior match numpy instead? /cc @tensorflow/api-owners

**Comment by alextp** [comment]
+1 to @yongtang 's comment

**Comment by mihaimaruseac** [comment]
+1. In Math there is no comparison between two complex numbers, so if we want to implement that support then we should make sure to follow existing practices

**Review by mihaimaruseac — COMMENTED** [review]
Let's also add some test where we compare several numbers. Scenarios I would include: 

1. Numbers with same real part
2. Numbers with same imaginary part
3. Numbers all around the complex space (maybe some roots of unity?)

**Comment by jstzwj** [comment]
Argmax returns the indices of the maximum values ignoring NaNs. 
In numpy, np.argmax return the first number that is NAN when there are nans in array, and np.nanargmax returns the indices of the maximum values ignore nan. Currently, the behavior of argmax in tensorflow ignores nans.
Should tensorflow be consistent with numpy at this point?

Numpy 1.18.1, Python3.8:
```python3
import numpy as np
print(np.argmax([complex(3,4), complex(1, np.nan), complex(np.nan, 2)]))
```
**Output**
```
1
```
However, the numpy documentation does not seem to mention it, this may be undefined behavior?

In addition , when I try to implement complex support for argmax and argmin, I found it is strange that `lhs_value` and `rhs_value` seem to be swapped in `tensorflow/compiler/xla/client/lib/arithmetic.cc:CreateMinMaxComputation`, when the input is complex type. This causes my argmax always choose the last nan if there are nans in tensor...

**Review by mihaimaruseac — DISMISSED** [review]

File: tensorflow/compiler/tests/argminmax_test.py [hunk 1]
```diff
@@ -93,7 +93,237 @@ def testArgMinMax(self):
             output_type=output_type,
             op_input=np.array([[4, 1], [3, 2]], dtype=dtype),
             expected=np.array([1, 1], dtype=output_type))
+  def testFloat(self):
+    # Test complex numbers support of argmin/argmax.
```
Comment (redwrasse) [inline tensorflow/compiler/tests/argminmax_test.py:97]: I think you mixed this up, not the comment intended for `testFloat`

File: tensorflow/compiler/tests/argminmax_test.py [hunk 2]
```diff
@@ -51,7 +51,7 @@ def _assertOpOutputMatchesExpected(self, op, axis, output_type, op_input,
   def testArgMinMax(self):
     # Complex numbers do not support argmin/argmax.
```
Comment (redwrasse) [inline tensorflow/compiler/tests/argminmax_test.py:52]: to be removed now?: ` # Complex numbers do not support argmin/argmax.`

**Comment by redwrasse** [comment]
> Let's also add some test where we compare several numbers. Scenarios I would include:
> 
> 1. Numbers with same real part
> 2. Numbers with same imaginary part
> 3. Numbers all around the complex space (maybe some roots of unity?)

It would be nice to have somewhere centrally in the Tensorflow documentation what the expected behavior on complex numbers is and why. That principle would then determined the behavior of all operations. For example this choice of using the real component would suggest preserve the 
elementary operations (addition, multiplication) but use the real component norm as norm function as a general principle for complex numbers. But there is at least one paper on 'complex neural networks', like this one: https://arxiv.org/abs/1705.09792. And poking around on stackoverflow some frustration about incomplete support for complex numbers https://stackoverflow.com/questions/43934487/minimizing-loss-in-a-complex-valued-network-in-tensorflow

**Comment by jstzwj** [comment]
I try to add test cases in `tensorflow\compiler\tests\argminmax_test.py`. However, it seems that there are some problems in the existing code, the following code cannot pass even in the master branch.

```python3
def testNaN(self):
    minmax_types = self.all_types & {np.int32, np.int64}
    for dtype in self.float_types:
      for output_type in minmax_types:
        # argmax NAN tests (ignore nan in tensor)
        self._assertOpOutputMatchesExpected(
            math_ops.argmax,
            axis=0,
            output_type=output_type,
            op_input=np.array([np.nan, 6.0, 7.0, -1.0, 4.0, np.nan, -50.0], dtype=dtype),
            expected=output_type(2))
```

Output
```
AssertionError:
Arrays are not equal

not equal lhs = array(6)
not equal rhs = array(2)
Mismatched elements: 1 / 1 (100%)
Max absolute difference: 4
Max relative difference: 2.
 x: array(6)
 y: array(2)

```

**Comment by sanjoy** [comment]
@agarwal-ashish any suggestions on what the semantic should be?

**Comment by gbaned** [comment]
@wangpengmit  Can you please take a look on this PR ? Thanks!

**Comment by wangpengmit** [comment]
> I try to add test cases in `tensorflow\compiler\tests\argminmax_test.py`. However, it seems that there are some problems in the existing code, the following code cannot pass even in the master branch.
> 
> ```python
> def testNaN(self):
>     minmax_types = self.all_types & {np.int32, np.int64}
>     for dtype in self.float_types:
>       for output_type in minmax_types:
>         # argmax NAN tests (ignore nan in tensor)
>         self._assertOpOutputMatchesExpected(
>             math_ops.argmax,
>             axis=0,
>             output_type=output_type,
>             op_input=np.array([np.nan, 6.0, 7.0, -1.0, 4.0, np.nan, -50.0], dtype=dtype),
>             expected=output_type(2))
> ```
> 
> Output
> 
> ```
> AssertionError:
> Arrays are not equal
> 
> not equal lhs = array(6)
> not equal rhs = array(2)
> Mismatched elements: 1 / 1 (100%)
> Max absolute difference: 4
> Max relative difference: 2.
>  x: array(6)
>  y: array(2)
> ```

Is it because `all_types` has unsigned integers?

**Review by wangpengmit — APPROVED** [review]

**Comment by gbaned** [comment]
@jstzwj  Can you please address Ubuntu Sanity errors? Thanks!

**Review by wangpengmit — APPROVED** [review]

**Comment by majnemer** [comment]
Thanks for making this contribution!

I think it might be nicer for the complex argmin/argmax to explicitly extract the real/complex components rather than defining minimum and maximum for complex values. I think this should result in the same performance but avoids having to define an ordering of complex elements at the HLO level.

**Comment by gbaned** [comment]
@jstzwj  Any update on this PR? Please. Thanks!

**Comment by cantonios** [comment]
I agree with @majnemer - unless we explicitly want to support the same semantics as numpy, it's probably best *not* to implement argmin/argmax for complex types, but instead force the user to be explicit about the semantics.

Defining argmin/argmax for complex will force us to also define these semantics elsewhere, for all backends (XLA-HLO, Eigen, etc...).  In Eigen we will not implement complex comparison functions like this, since there is no mathematical basis for it - so TF will need to maintain its own internal workaround.

**Comment by gbaned** [comment]
@jstzwj Any update on this PR? Please. Thanks!

**Review by mihaimaruseac — CHANGES_REQUESTED** [review]
(leaving a request changes so that this does not get imported again until the previous comments are handled)

**Comment by gbaned** [comment]
@jstzwj Can you please resolve conflicts? Thanks!

**Comment by gbaned** [comment]
@jstzwj  Any update on this PR? Please. Thanks!

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by gbaned** [comment]
@jstzwj Any update on this PR? Please. Thanks!

**Comment by tensorflowbutler** [comment]
It has been 28 days that this pull-request has stalled. Please create a new pull-request with the requested changes.
